// @snekcode/mui-search-bar/lib/components/SearchBar/SearchBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _Input = _interopRequireDefault(require("@material-ui/core/Input"));

var _Paper = _interopRequireDefault(require("@material-ui/core/Paper"));

var _Clear = _interopRequireDefault(require("@material-ui/icons/Clear"));

var _Search = _interopRequireDefault(require("@material-ui/icons/Search"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _classnames = _interopRequireDefault(require("classnames"));

var _excluded = ["cancelOnEscape", "className", "classes", "closeIcon", "disabled", "onCancelSearch", "onRequestSearch", "searchIcon", "style", "placeholder", "value"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

//@ts-ignore
var styles = function styles(theme) {
  return {
    root: {
      height: theme.spacing(6),
      display: "flex",
      justifyContent: "space-between"
    },
    iconButton: {
      color: theme.palette.action.active,
      transform: "scale(1, 1)",
      transition: theme.transitions.create(["transform", "color"], {
        duration: theme.transitions.duration.shorter,
        easing: theme.transitions.easing.easeInOut
      })
    },
    iconButtonHidden: {
      transform: "scale(0, 0)",
      "& > $icon": {
        opacity: 0
      }
    },
    searchIconButton: {
      marginRight: theme.spacing(-6)
    },
    icon: {
      transition: theme.transitions.create(["opacity"], {
        duration: theme.transitions.duration.shorter,
        easing: theme.transitions.easing.easeInOut
      })
    },
    input: {
      width: "100%"
    },
    searchContainer: {
      margin: "auto 16px",
      width: "calc(100% - ".concat(theme.spacing(6 + 4), "px)") // 6 button + 4 margin

    }
  };
};

/**
 * Material design search bar
 * @see [Search patterns](https://material.io/archive/guidelines/patterns/search.html)
 */
var SearchBar = /*#__PURE__*/_react["default"].forwardRef(function (_ref, ref) {
  var cancelOnEscape = _ref.cancelOnEscape,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? "" : _ref$className,
      classes = _ref.classes,
      _ref$closeIcon = _ref.closeIcon,
      closeIcon = _ref$closeIcon === void 0 ? /*#__PURE__*/_react["default"].createElement(_Clear["default"], null) : _ref$closeIcon,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      onCancelSearch = _ref.onCancelSearch,
      onRequestSearch = _ref.onRequestSearch,
      _ref$searchIcon = _ref.searchIcon,
      searchIcon = _ref$searchIcon === void 0 ? /*#__PURE__*/_react["default"].createElement(_Search["default"], null) : _ref$searchIcon,
      _ref$style = _ref.style,
      style = _ref$style === void 0 ? undefined : _ref$style,
      _ref$placeholder = _ref.placeholder,
      placeholder = _ref$placeholder === void 0 ? "Search" : _ref$placeholder,
      userDefinedValue = _ref.value,
      inputProps = _objectWithoutProperties(_ref, _excluded);

  var inputRef = _react["default"].useRef({});

  var _React$useState = _react["default"].useState(userDefinedValue),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  _react["default"].useEffect(function () {
    setValue(userDefinedValue);
  }, [userDefinedValue]);

  var handleFocus = _react["default"].useCallback(function (e) {
    if (inputProps.onFocus) {
      inputProps.onFocus(e);
    }
  }, [inputProps.onFocus]);

  var handleBlur = _react["default"].useCallback(function (e) {
    setValue(function (v) {
      return v.trim();
    });

    if (inputProps.onBlur) {
      inputProps.onBlur(e);
    }
  }, [inputProps.onBlur]);

  var handleInput = _react["default"].useCallback(function (e) {
    setValue(e.target.value);
    inputProps.onChange(e.target.value);
  }, [inputProps.onChange]);

  var handleCancel = _react["default"].useCallback(function () {
    setValue("");
    onCancelSearch();
  }, [onCancelSearch]);

  var handleRequestSearch = _react["default"].useCallback(function () {
    if (onRequestSearch) {
      onRequestSearch(value);
    }
  }, [onRequestSearch, value]);

  var handleKeyUp = _react["default"].useCallback(function (e) {
    if (e.charCode === 13 || e.key === "Enter") {
      handleRequestSearch();
    } else if (cancelOnEscape && (e.charCode === 27 || e.key === "Escape")) {
      handleCancel();
    }

    if (inputProps.onKeyUp) {
      inputProps.onKeyUp(e);
    }
  }, [handleRequestSearch, cancelOnEscape, handleCancel, inputProps.onKeyUp]);

  return /*#__PURE__*/_react["default"].createElement(_Paper["default"], {
    className: (0, _classnames["default"])(classes === null || classes === void 0 ? void 0 : classes.root, className),
    style: style
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classes === null || classes === void 0 ? void 0 : classes.searchContainer
  }, /*#__PURE__*/_react["default"].createElement(_Input["default"], _extends({}, inputProps, {
    inputRef: inputRef,
    onBlur: handleBlur,
    value: value,
    onChange: handleInput,
    onKeyUp: handleKeyUp,
    onFocus: handleFocus,
    fullWidth: true,
    className: classes === null || classes === void 0 ? void 0 : classes.input,
    disableUnderline: true,
    disabled: disabled,
    placeholder: placeholder
  }))), /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    role: "searchButton",
    onClick: handleRequestSearch,
    className: (0, _classnames["default"])(classes === null || classes === void 0 ? void 0 : classes.iconButton, classes === null || classes === void 0 ? void 0 : classes.searchIconButton // {
    //   [classes?.iconButtonHidden as string]: value !== "",
    // }
    ),
    disabled: disabled
  }, /*#__PURE__*/_react["default"].cloneElement(searchIcon, {
    classes: {
      root: classes === null || classes === void 0 ? void 0 : classes.icon
    }
  })));
});

var _default = (0, _withStyles["default"])(styles)(SearchBar);

exports["default"] = _default;
